### PR TITLE
Imports: increase point penalty for non-matching publishers

### LINKED ISSUE #9387: ImportBot ignoring publishers field when resolving editions
### Problem

Related: #8977 

It seems to think a title match is enough to consider something a match :/

E.g. https://openlibrary.org/books/OL46219062M/Spoon_River_Anthology?m=diff&b=2

Note I have since undone this import.

### Reproducing the bug

1. copydocs this record into your local environment: https://openlibrary.org/books/OL46219062M/Spoon_River_Anthology
1. Try to import this record:

```json
{
    "title": "Spoon River Anthology",
    "source_records": ["standard_ebooks:edgar-lee-masters/spoon-river-anthology"],
    "publishers": ["Standard Ebooks"],
    "publish_date": "2022",
    "authors": [{"name": "Edgar Lee Masters"}],
    "description": "<p><i>Spoon River Anthology</i> is a collection of short poems that reveals the true nature of the citizens of a fictional small town in Illinois. Each poem is a candid autobiography of a now-deceased resident that lies in the Oak Hill Cemetery\u2014often exposing their darkest secrets.</p> <p><a href=\"https://standardebooks.org/ebooks/edgar-lee-masters\">Edgar Lee Masters</a> was raised in Lewistown, Illinois and based these stories on the gossip he heard there. The book was a commercial success, but was banned from schools and libraries in the area due to the real-life citizens knowing exactly whom each poem was written about. As the years have passed and more generations now lie in Oak Hill Cemetery, Lewistown has forgiven Masters, and he\u2019s now celebrated there.</p> <p>The poems were originally published in the literary magazine <i>Reedy\u2019s Mirror</i> under the pseudonym Webster Ford. The first book edition was published in 1915 and contained 209 poems. Masters added 35 new poems, including the epilogue, for the 1916 edition, which is the edition that this Standard Ebooks edition is based on.</p>",
    "subjects": ["American poetry", "Poetry"],
    "identifiers": {
        "standard_ebooks": ["edgar-lee-masters/spoon-river-anthology"]
    },
    "languages": ["eng"],
    "cover": "https://standardebooks.org/ebooks/edgar-lee-masters/spoon-river-anthology/downloads/cover.jpg"
}
```

* Expected behavior: A new edition is created.
* Actual behavior: It's attached to this edition!

### Context

- Browser (Chrome, Safari, Firefox, etc):
- OS (Windows, Mac, etc):
- Logged in (Y/N): Y
- Environment (prod, dev, local): prod

#### Stakeholders
@scottbarnes 

**Note:** Before [making a new branch](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#making-changes-and-creating-a-pull-request) _or_ updating an existing one, please [ensure your branch is up to date](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#working-on-your-branch).

### PR DESCRIPTION
**Pull request by scottbarnes**
Closes #9387 

Fix.

This PR increases the penalty for a truthy and yet non-matching `publishers` field so that during import editions will not match when they have:
1. matching author;
2. matching title;
3. matching publish_date; and
4. do NOT have a matching truthy publishers field.

Edition matching uses a points system found in
`openlibrary/catalog/add_book/match.py`, and a match is made when the threshold of 875 points is met. This PR increases the penalty for non-matching publishers to stop a match from occurring in the above situation.

The rationale is that multiple editions can be published by different publishers in the same year, and all would have the same title, authors, and publish_date, yet they would be different editions if the truthy `publishers` field did not match.

If the `publishers` field is falsy, then a match should happen as it would before, though this does create the risk that the incoming edition could have an ISBN or other identifier that doesn't match the existing publisher from the original record; however, this was the case before and hitherto that has not appeared to be an issue.

### Technical

For reasons I have not yet determined, the `status` field is coming back `modified` for the Work when importing the record suggested in #9387. However, the Work record is itself not modified, which makes sense, as the code changes were to functions called by `editions_match`. Still, this is puzzling and I'm not yet sure if it's an artifact of `copydocs.py` import, or perhaps the fact that I've imported and deleted these records a few times, so their revision history is a bit odd.

### Testing

Import the records as mentioned in #9387:
```
docker compose exec web bash
library@9f99aad1aca5:/openlibrary$ PYTHONPATH=. ./scripts/copydocs.py /works/OL31830589W?v=2 /authors/OL7636796A?v=1 /books/OL46219062M?v=3
fetching versioned [KeyVersionPair(key='/works/OL31830589W', version='2'), KeyVersionPair(key='/authors/OL7636796A', version='1'), KeyVersionPair(key='/books/OL46219062M', version='3')]
saving ['/works/OL31830589W', '/authors/OL7636796A', '/books/OL46219062M']
[{'key': '/works/OL31830589W', 'revision': 6}, {'key': '/authors/OL7636796A', 'revision': 3}, {'key': '/books/OL46219062M', 'revision': 11}]
```
```
❯ curl -X POST \
     -H "Content-Type: application/json" \
     -b ~/cookies.txt \
     -d '{
    "title": "Spoon River Anthology",
    "source_records": ["standard_ebooks:edgar-lee-masters/spoon-river-anthology"],
    "publishers": ["Standard Ebooks"],
    "publish_date": "2022",
    "authors": [{"name": "Edgar Lee Masters"}],
    "description": "<p><i>Spoon River Anthology</i> is a collection of short poems that reveals the true nature of the citizens of a fictional small town in Illinois. Each poem is a candid autobiography of a now-deceased resident that lies in the Oak Hill Cemetery\u2014often exposing their darkest secrets.</p> <p><a href=\"https://standardebooks.org/ebooks/edgar-lee-masters\">Edgar Lee Masters</a> was raised in Lewistown, Illinois and based these stories on the gossip he heard there. The book was a commercial success, but was banned from schools and libraries in the area due to the real-life citizens knowing exactly whom each poem was written about. As the years have passed and more generations now lie in Oak Hill Cemetery, Lewistown has forgiven Masters, and he\u2019s now celebrated there.</p> <p>The poems were originally published in the literary magazine <i>Reedy\u2019s Mirror</i> under the pseudonym Webster Ford. The first book edition was published in 1915 and contained 209 poems. Masters added 35 new poems, including the epilogue, for the 1916 edition, which is the edition that this Standard Ebooks edition is based on.</p>",
    "subjects": ["American poetry", "Poetry"],
    "identifiers": {
        "standard_ebooks": ["edgar-lee-masters/spoon-river-anthology"]
    },
    "languages": ["eng"],
    "cover": "https://standardebooks.org/ebooks/edgar-lee-masters/spoon-river-anthology/downloads/cover.jpg"
     }' \
     http://localhost:8080/api/import

{"authors": [{"key": "/authors/OL7636796A", "name": "Edgar Lee Masters", "status": "matched"}], "success": true, "edition": {"key": "/books/OL14M", "status": "created"}, "work": {"key": "/works/OL31830589W", "status": "modified"}}
```

### Screenshot

The new, matched, edition:
image

### Stakeholders

@hornc, I'd especially appreciate any feedback you have.
@cdrini

### CODE REVIEW

**Review by cdrini — APPROVED** [review]
This looks good to me; it's hard to verify code like this since it's vaguely neural-network-y/ML-y, but we tried a few examples and it seemed to behave roughly correctly. I think it makes sense that a publisher mismatch should be a BIG red flag for edition matching.

**Comment by cdrini** [comment]
@hornc Would love if you good give a brief 👀 to make sure this looks kosher!

**Comment by tfmorris** [comment]
This looks like an improvement to me, but I agree with Drini that this 1990s era classifier with hand tuned weights and thresholds could use serious improvement to make it easier to review. Ideally  it would be replaced with an actual ML classifier fed with real world training data and ground truths.

On the feature engineering front, regularizing and deduping the normalization processes would make things more concise and easer to review. In addition to the already noted TODO to expand article elision to non-English languages, it should also remove title words which are subject to cataloging variability such as "novel," "roman," "roman policier," etc.

**Comment by hornc** [comment]
Some general comments on this feature, before and after the changes:

1. I agree the weighting system is a bit opaque. More test cases with justifications would be helpful in general.
2. This change is probably going to have the effect that variations on a publisher's name will now create new records which will need to be merged. Publishers are not controlled ids, and publisher strings do have more of a range of formats, abbreviations which can refer to the same publisher in catalogs.
3. Further to 2., I think this was a _deliberate feature_ of OL, and the assumption that older books with title and author from the same year are the same edition (and therefore "published" by the same source) isn't a terrible one in practice.
4. The one use/test case example provided here isn't explained in detail, and TBH in reviewing it I question whether this isn't a self-published ebook distributed through multiple platforms or sources, so is combining them really incorrect? Are they different editions, is the content different, was the content created and laid out by the same people? 

I don't really know on 4, now that I've looked harder it seems this is a historical reprint rather than self-published poetry, but it looks like Creative Media Partners LLC is a republisher of open books, so I don't know whether they are resusing the Standard ebooks content? It seems like that would be a legitimate possibility given this is all open.

I think this is trading off one heuristic that mostly worked with traditional publishing for one that might work better for open ebook publishers, without making motivations or distinctions clear in a one-size-fits-all way. My gut feel is the old assumption is still more applicable, but the import behavior will change and it'll be worked around by librarian edits to fix anything that looks wrong.

Main points arising from this:
* Publisher string matching is more complex, and has less payoff than title/author matching, which is why it was avoided previously. This PR also sidesteps the challenge, and it will likely cause more duplicate records, which may not be the desired effect.
* What are OL's policies for Open and other ebook publishing? The nature of ebook publishing changes some basic assumptions about publishing, and I'm not sure there are clear guidelines on we should handle platforms for self-publishing. I'm sure I've see records which claim a book is published by Amazon. There is a lot of potential for noise anyway.
* It feels like this is a different and new ebook specific usecase from what works for library cataloged material. Would a library catalog the example items in two bib records? (Honest question; I don't know, I think potentially not, but some up-to-date guidance and reasoning would be good)

I believe Standard ebooks may be a trusted partner? Perhaps there is a special publisher source import usecase which can rely on more structured publisher metadata to permit strict publisher string matching (via a mode / import parameter), but the general library bib-record import can still use the older non-strict matching?

**Comment by hornc** [comment]
This also potentially relates to print-on-demand generated catalogs. Here's a list of this title with publishers like 'digireads.com', 'lulu.com' and Createspace: https://www.libroworld.com/search.php?search_query=%22Spoon+River+Anthology%22  Many have ISBNs and are potentially print books in addition to ebooks, if someone ever orders one. archive.org has scanned POD reprints like this. What a time we are living in...

**Comment by cdrini** [comment]
Awesome thank you @hornc ! Yeah, that was my concern as well that we would now be creating unnecessary editions. Note we did also manually test a case where the publisher was a substring of another publisher ; that caused it to bump up enough to modify the existing edition.

Yep! Standard Ebooks is a [trusted book provider](https://blog.openlibrary.org/2021/12/20/introducing-trusted-book-providers/); they publish ebooks of classics, so they should always be their own, independent editions.

> What are OL's policies for Open and other ebook publishing?

I think evolving :P Trusted Book Providers are examples of a place we want to link to the ebooks online though!

> Would a library catalog the example items in two bib records?

I believe so; with the covers being different at the very least, my guess would be these two would be considered separate editions. Also since both have at the very least different stated publishers! Even the CreateSpace one, I believe has the wrong publisher listed. Other books published independently often list a different publisher (indeed on Amazon it's listed as Legare Street Press!). Also a lot of these public domain independent publishing groups tend to also publish facsimiles of older editions, or potentially will use the Project Gutenberg epub, so without further evidence, in this case, I don't think the year/title are enough to qualify matching.

But I agree with you in general for older books it might cause more duplicate editions to be created based on small publisher name variations :/ Hopefully the substring check will help catch some of those, but I think the best approach might be to run with this, and if we spot examples of bad imports, then we can capture them and encode them as unit tests, and update the model. That way we can make sure they get fixed and stay fixed! (Who knows, we could potentially make this behaviour dependent on the year in the future if we find that more accurately segments the data!)

I'll post on slack warning librarians to keep an eye out!

**Comment by cdrini** [comment]
This is a bit of a murky space, but I would call these as separate publishers; Standard Ebooks edits the text, removes typos, and reformats the text as well. IA's are more clear mirrors of the original edition. Similarly with Project Gutenberg, which has groups of editors that contribute to transforming the source edition into a new medium. Effectively retyping it (different from a facsimile). Not to mention they also have new covers for the works, which is extra metadata which can now become contradictory if attached to the same edition.

Hmmm I think maybe a useful metre stick is contradicting metadata. Which cover? Which publisher, Standard Ebooks or "CreateSpace"? Which year? Should we add a "Cover Artist" field? Which cover does it pertain to? I think in the case of Standard Ebooks, Project Gutenberg, there's enough contradicting data that leads me to think they should be separate editions. 

I agree that erring on the side of multiple editions is better though; merging is easy, conflating is hard.

**Comment by hornc** [comment]
Many print editions have different cover variations and should not have different records. OL supports this by modelling cover as a list. The UI used to support this more clearly, but I just checked this and it's considerably less clear now with the blurring of work and edition interfaces. Anyway, there's a saying about judging books by covers ;) 

Joking aside, I think covers are a bad metric for a new edition,  _especially_ with POD / unscrupulous re-printers. I accept that Standard Ebooks appears very different, but they are so different the cases should be differentiated. 

Also, FWIW, I don't consider Createspace to be a publisher, and I don't even mean that to disparage Createspace (or anyone who publishes via them). I'd class those books as self-published, and it might be more useful to specify the individual responsible, which I'd try to do if I were adding one. There seems to be no guidance though, so I don't expect any consistency. 

Bibliographic cataloging is really about differentiated content. The rabbit hole of reprints goes quite deep -- (mutliple) reprinters with PD digital catalogs offer leather bound copies of anything you like on Amazon and other places, I didn't even have to look hard, but guessed there'd be a copy of the book we're talking about, and  here it is: https://www.amazon.com/Spoon-River-anthology-Leather-Bound/dp/B0B5T39V76   It even has a positive review. I think these things are close to scams, or at best like buying an upholstered household decoration rather than a book. If these are treated as real publishers, we're going to have an explosion of records and noise.

I can't help feeling the Standard Ebooks issue could be handled differently using the current publisher matching, and this feature will have more impact in increasing unscrupulous / noise editions.

 I worry that the legitimate PD ebooks will be used to multiply these un-differentiated money grabs too.
